torch/CMakeLists.txt

```diff
@@ -265,73 +183,145 @@ if (NOT NO_API)
     ${TORCH_SRC_DIR}/csrc/api/src/optim/lbfgs.cpp
     ${TORCH_SRC_DIR}/csrc/api/src/optim/rmsprop.cpp
     ${TORCH_SRC_DIR}/csrc/api/src/optim/sgd.cpp
-  )
+    )
+
 endif()
 
 add_library(torch SHARED ${TORCH_SRCS})
 
-# https://gcc.gnu.org/onlinedocs/gcc-4.0.3/gcc/Warning-Options.html
-target_compile_options(torch
-  PRIVATE
-  -Wall
-  -Wextra
-  -pedantic
-  -Wcast-align
-  -Wcast-qual
-  -Wctor-dtor-privacy
-  -Wdisabled-optimization
-  -Winit-self
-  -Wmissing-include-dirs
-  -Woverloaded-virtual
-  -Wsign-promo
-  -Wstrict-overflow=5
-  -Wundef
-  -fdiagnostics-show-option
-  -Wno-unused-parameter
-  -Wno-missing-braces # This warning is buggy
-  -Wno-unknown-pragmas)
-
-if ($ENV{WERROR})
+# until they can be unified, keep these lists synced with setup.py
+if(MSVC)
+  target_compile_options(torch PRIVATE
+    /MD
+    /Z7
+    /EHa
+    /DNOMINMAX
+    /wd4267
+    /wd4251
+    /wd4522
+    /wd4522
+    /wd4838
+    /wd4305
+    /wd4244
+    /wd4190
+    /wd4101
+    /wd4996
+    /wd4275
+    /bigobj
+    )
+else()
+  target_compile_options(torch PRIVATE
+    -std=c++11
+    -Wall
+    -Wextra
+    -Wno-unused-parameter
+    -Wno-missing-field-initializers
+    -Wno-write-strings
+#    -Wno-zero-length-array
+    -Wno-unknown-pragmas
+    # This is required for Python 2 declarations that are deprecated
+    # in 3.
+    -Wno-deprecated-declarations
+    # Python 2.6 requires -fno-strict-aliasing see
```

Comment:
I'm curious if we couldn't have some flag to know whether this build is for Python, or not for Python? If not for Python, we wouldn't need this `no-strict-aliasing` flag for example.

Reply:
Regarding all of the various warnings flags - I haven't introduced any code constructs that elicit new warnings and/or errors. They're all needed by pre-existing code, and the Werror violations are exposed by the fact that all compliations/build environments are now driven through this unified codepath.

I claim that it's slightly out-of-scope of this diff to fix all the underlying, pre-existing errors, beyond what's needed to get a green CI. I'd rather record this in the unification-code-debt list and churn through them in a follow-up diff.

Reply:
What are the new "compilations"/"build environments"? Just to understand

Reply:
for example, the fact that torch is built by cmake on windows - this was never the case before. Or, test_api being compiled with the set of compiler flags taken from setup.py